nearcore/src/state_sync.rs

```diff
@@ -128,34 +132,69 @@ impl StateSyncDumpHandle {
     }
 }
 
-/// A thread loop that dumps state of the latest available epoch to S3.
-/// Operates as a state machine. Persists its state in the Misc column.
-/// Shards store the state-machine state separately.
+fn extract_part_id_from_part_file_name(file_name: &String) -> u64 {
+    let re = Regex::new(r"state_part_[0-9]+_of_[0-9]+$").unwrap();
+    assert!(re.is_match(file_name));
+    let splitted: Vec<_> = file_name.split("_").collect();
+    let part_id = splitted.get(2).unwrap().to_string().parse::<u64>().unwrap();
+    return part_id;
+}
+
+async fn get_missing_state_parts_for_epoch(
+    shard_id: ShardId,
+    chain_id: &String,
+    epoch_id: &EpochId,
+    epoch_height: u64,
+    total_parts: u64,
+    external: &ExternalConnection,
+) -> Result<Vec<u64>, anyhow::Error> {
+    let directory_path =
+        external_storage_location_directory(chain_id, epoch_id, epoch_height, shard_id);
+    let file_names = external.list_state_parts(shard_id, &directory_path).await?;
+    if !file_names.is_empty() {
+        let existing_nums: HashSet<_> = file_names
+            .iter()
+            .map(|file_name| extract_part_id_from_part_file_name(file_name))
+            .collect();
+        let missing_nums: Vec<u64> =
+            (0..total_parts).filter(|i| !existing_nums.contains(i)).collect();
+        let num_missing = missing_nums.len();
+        tracing::debug!(target: "state_sync_dump", ?num_missing, ?directory_path, "number of missing parts: ");
+        Ok(missing_nums)
+    } else {
+        tracing::debug!(target: "state_sync_dump", ?total_parts, ?directory_path, "number of missing parts: ");
+        let missing_nums = (0..total_parts).collect::<Vec<_>>();
+        Ok(missing_nums)
+    }
+}
+
+fn select_random_part_id_with_index(parts_to_be_dumped: &Vec<u64>) -> (u64, usize) {
+    let mut rng = thread_rng();
+    let selected_idx = rng.gen_range(0..parts_to_be_dumped.len());
+    let selected_element = parts_to_be_dumped[selected_idx];
+    tracing::debug!(target: "state_sync_dump", ?selected_element, "selected parts to dump: ");
+    (selected_element, selected_idx)
+}
+
 async fn state_sync_dump(
     shard_id: ShardId,
     chain: Chain,
     epoch_manager: Arc<dyn EpochManagerAdapter>,
     shard_tracker: ShardTracker,
     runtime: Arc<dyn RuntimeAdapter>,
     chain_id: String,
-    restart_dump_for_shards: Vec<ShardId>,
```

Comment:
Please consider removing `restart_dump_for_shards` from `ClientConfig` and from `Config` (i.e. `NearConfig::config`)